native_client/beam_search.cc

```diff
@@ -0,0 +1,546 @@
+      s = decoded_values->allocate(p, tf::TensorShape({p_num}), &p_values);
+      if (!s.ok()) return s;
+      s = decoded_shape->allocate(p, tf::TensorShape({2}), &p_shape);
+      if (!s.ok()) return s;
+
+      auto indices_t = p_indices->matrix<tf::int64>();
+      auto values_t = p_values->vec<tf::int64>();
+      auto shape_t = p_shape->vec<tf::int64>();
+
+      tf::int64 max_decoded = 0;
+      tf::int64 offset = 0;
+
+      for (tf::int64 b = 0; b < batch_size; ++b) {
+        auto &p_batch = sequences[b][p];
+        tf::int64 num_decoded = p_batch.size();
+        max_decoded = std::max(max_decoded, num_decoded);
+        std::copy_n(p_batch.begin(), num_decoded, &values_t(offset));
+        for (tf::int64 t = 0; t < num_decoded; ++t, ++offset) {
+          indices_t(offset, 0) = b;
+          indices_t(offset, 1) = t;
+        }
+      }
+
+      shape_t(0) = batch_size;
+      shape_t(1) = max_decoded;
+    }
+    return tf::Status::OK();
+  }
+
+ private:
+  int top_paths_;
+  TF_DISALLOW_COPY_AND_ASSIGN(CTCDecodeHelper);
+};
+
+// CTC beam search
+class CTCBeamSearchDecoderOp : public tf::OpKernel {
+ public:
+  explicit CTCBeamSearchDecoderOp(tf::OpKernelConstruction *ctx)
+    : tf::OpKernel(ctx)
+    , beam_scorer_(GetModelPath(ctx),
+                   GetTriePath(ctx),
+                   GetAlphabetPath(ctx))
+  {
+    OP_REQUIRES_OK(ctx, ctx->GetAttr("merge_repeated", &merge_repeated_));
+    OP_REQUIRES_OK(ctx, ctx->GetAttr("beam_width", &beam_width_));
+    int top_paths;
+    OP_REQUIRES_OK(ctx, ctx->GetAttr("top_paths", &top_paths));
+    decode_helper_.SetTopPaths(top_paths);
+
+    // const tf::Tensor* model_tensor;
+    // tf::Status status = ctx->input("model_path", &model_tensor);
+    // if (!status.ok()) return status;
+    // auto model_vec = model_tensor->flat<std::string>();
+    // *model_path = model_vec(0);
+
+    // const tf::Tensor* trie_tensor;
+    // status = ctx->input("trie_path", &trie_tensor);
+    // if (!status.ok()) return status;
+    // auto trie_vec = trie_tensor->flat<std::string>();
+    // *trie_path = model_vec(0);
+
+    // const tf::Tensor* alphabet_tensor;
+    // status = ctx->input("alphabet_path", &alphabet_tensor);
+    // if (!status.ok()) return status;
+    // auto alphabet_vec = alphabet_tensor->flat<std::string>();
+    // *alphabet_path = alphabet_vec(0);
+  }
+
+  std::string GetModelPath(tf::OpKernelConstruction *ctx) {
+    std::string model_path;
+    ctx->GetAttr("model_path", &model_path);
+    return model_path;
+  }
+
+  std::string GetTriePath(tf::OpKernelConstruction *ctx) {
+    std::string trie_path;
+    ctx->GetAttr("trie_path", &trie_path);
+    return trie_path;
+  }
+
+  std::string GetAlphabetPath(tf::OpKernelConstruction *ctx) {
+    std::string alphabet_path;
+    ctx->GetAttr("alphabet_path", &alphabet_path);
+    return alphabet_path;
+  }
+
+  void Compute(tf::OpKernelContext *ctx) override {
+    const tf::Tensor *inputs;
```

Comment:
Why such C based C++? Introducing all the variables at the start of the function.

Reply:
This code is copied from `CTCBeamSearchDecoder` in tensorflow, mostly unmodified. All I change is which beam scorer to use. I can make the code more C++-y, but then if we ever need to uplift changes from there to here it's harder.

Reply:
OK, understood.